### PR TITLE
Dynamically generate types for unseen data, and fix GitHub integration

### PR DESCRIPTION
**Pull request by Alfred-Mountfield**
## 🌟 What is the purpose of this PR?

Quite a while ago we added a preliminary "integration" that was able to ingest data from GitHub (#503). This was able to be ran in development environments, but wasn't part of CI or our deployments.

That development work helped support the need for the development of the Block Protocol Type System, and work was paused on it. This PR represents a small return to the work of integrations, and introduces dynamic generation of types (expressed within the Type System) for unseen JSON data. Utilising this work, it "fixes" the GitHub integration, under the state it was originally introduced in; meaning it will still not be available in deployments, and isn't ready for usage outside of development.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203363157432094/1203657039764067/f) _(internal)_
- [Asana task](https://app.asana.com/0/1203363157432094/1203657039764066/f) _(internal)_

## 🚫 Blocked by

N/A

## 🔍 What does this change?

Please see commits

## 📜 Does this require a change to the docs?

- As this is reserved for development, it is currently not well captured in docs. This will be fixed in the future as the code becomes a more stable part of the codebase.

## ⚠️ Known issues

- Running multiple ingests in a row for the same data streams will likely result in unwanted behaviour (potentially crashes), we currently do not check for existing data (types or entities) before trying to create them
- Ingest is very slow, this is partially due to the nature of the Airbyte connectors, but also due to inefficient code written here (again, just for development purposes at the moment) and the lack of a bulk/batch insertion endpoint on the Graph (which means we need a roundabout call for every record)

## 🐾 Next steps

- Implement an Asana connector

## 🛡 What tests cover this?

- None besides lints

## ❓ How to test this?

1. Checkout the branch / view the deployment
1. Follow the README to start the app as normal (including `external-services` and `yarn dev`)
1. Run `yarn workspace @hashintel/hash-task-executor start`
1. Open the front-end and login to generate the appropriate cookies
1. Open the GraphQL playground (http://localhost:5001/graphql)
1. Acquire a [personal access token for GitHub](https://github.com/settings/tokens), the scope will depend on the streams you want to ingest, but I have attached my scopes at the bottom
1. Try the following mutation
   ```graphql
   mutation {
     executeGithubSpecTask
   }
   ```
1. Then you can try the following (replace the repositories, access token, and date as needed)
   ```graphql
   mutation {
     executeGithubCheckTask(config:{
       repository: "blockprotocol/blockprotocol hashintel/hash"
       start_date: "2023-01-01T00:00:00Z"
       credentials: {
         personal_access_token: ${PERSONAL_ACCESS_TOKEN}
       }
     })
   }
   ```
1. Then you can run the following to download the "catalog" for the connector (a list of available streams)
   ```graphql
   mutation {
     executeGithubDiscoverTask(config:{
       repository: "blockprotocol/blockprotocol hashintel/hash"
       start_date: "2023-01-01T00:00:00Z"
       credentials: {
         personal_access_token: ${PERSONAL_ACCESS_TOKEN}
       }
     })
   }
   ```
1. The previous mutation should have created a file: `packages/hash/task-executor/src/tasks/source-github/secrets/catalog.json`. Format the file using your IDE, and select a few "streams" that you want to download (delete the ones you do not want). The more you pick, the slower it will be to test. Some streams are not filterable by `start_date` so may take a long time even if you pick a recent `start_date`
1. Once you have modified `catalog.json` as you wish, run the following mutation (again updating parameters as you wish) 
   ```graphql
   mutation {
     executeGithubReadTask(config:{
       repository: "blockprotocol/blockprotocol hashintel/hash"
       start_date: "2023-01-01T00:00:00Z"
       credentials: {
         personal_access_token: ${PERSONAL_ACCESS_TOKEN}
       }
     })
   }
   ```
1. Check the app, you should see new entity types, and going to the relevant entity type page should show entities available.

## 📹 Demo
image
image
image

## Access Token Scopes
image
image

### CODE REVIEW

File: infra/terraform/hash/prod-usea1.tfvars [hunk 1]
```diff
@@ -28,7 +28,7 @@ kratos_env_vars = [
 ]
 
 hash_graph_env_vars = [
-  { name = "HASH_GRAPH_ALLOWED_URL_DOMAIN_PATTERN", secret = false, value = "https://alpha.hash.ai/@(?P<shortname>[\\w-]+)/types/(?P<kind>(?:data-type)|(?:property-type)|(?:entity-type))/[\\w-]+/" },
+  { name = "HASH_GRAPH_ALLOWED_URL_DOMAIN_PATTERN", secret = false, value = "https://alpha.hash.ai/@(?P<shortname>[\\w-]+)/types/(?P<kind>(?:data-type)|(?:property-type)|(?:entity-type))/[\\w\\-_%]+/" },
```
Comment (Alfred-Mountfield) [inline infra/terraform/hash/prod-usea1.tfvars:31]: Some of the keys have weird characters like `+`. I hope by URL encoding them we only need to add support for these characters, we'll find out with time I guess

File: packages/hash/external-services/docker-compose.yml [hunk 2]
```diff
@@ -120,7 +120,7 @@ services:
       HASH_GRAPH_PG_PORT: "5432"
       HASH_GRAPH_PG_DATABASE: "${HASH_GRAPH_PG_DEV_DATABASE}"
       HASH_GRAPH_LOG_FORMAT: "${HASH_GRAPH_LOG_FORMAT:-pretty}"
-      HASH_GRAPH_OTLP_ENDPOINT: "http://telemetry-collector:4317"
+#      HASH_GRAPH_OTLP_ENDPOINT: "http://telemetry-collector:4317"
```
Comment (Alfred-Mountfield) [inline packages/hash/external-services/docker-compose.yml:123]: This is a drive-by. The bug we discovered earlier meant that when I had an invalid domain URL we just didn't get any errors, the Graph deadlocked itself

Comment (Alfred-Mountfield) [reply packages/hash/external-services/docker-compose.yml:123]: This is tracked in https://app.asana.com/0/1201095311341924/1203636955054323/f

File: packages/hash/api/src/graphql/resolvers/task-executor/generation.ts [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+
+  return {
+    createdPropertyTypes,
+    createdEntityTypes,
+  };
+};
+
+/* eslint-disable no-param-reassign -- We want to mutate in place for efficiency */
+
+/// post-order traversal of the JSON object tree
```
Comment (thehabbos007) [inline packages/hash/api/src/graphql/resolvers/task-executor/generation.ts:284]: Comment seems to be superfluous

File: packages/hash/api/src/graphql/resolvers/task-executor/index.ts [hunk 4]
```diff
@@ -172,45 +149,145 @@ export const executeGithubReadTask: ResolverFn<
...
     );
-  } else {
-    const createdEntities = [];
-    try {
-      const airbyteRecords: AirbyteRecords = await taskExecutor.runTask(
-        Task.GithubRead,
-        config,
+  }
+
+  if (!user.shortname) {
```
Comment (thehabbos007) [inline packages/hash/api/src/graphql/resolvers/task-executor/index.ts:154]: I thought the `loggedInAndSignedUpMiddleware` took care of ensuring these values exist. @benwerner01 is there an alternative way for us to supply the user context in the case of the middleware that knows the shortname exists when the user has completed signup?

File: packages/hash/api/src/graphql/resolvers/task-executor/index.ts [hunk 5]
```diff
@@ -172,45 +149,145 @@ export const executeGithubReadTask: ResolverFn<
...
+      streamToEntityTypes[streamName] = updatedEntityType;
     }
+
+    logger.debug(
+      "Finished rewriting data into entities and types, inserting into the Graph.",
+    );
+
+    const propertyTypeMap: Record<
+      VersionedUri,
+      { schema: PropertyType; created: boolean }
```
Comment (thehabbos007) [inline packages/hash/api/src/graphql/resolvers/task-executor/index.ts:218]: The `created` property here and below seem to be unused?

File: packages/hash/external-services/docker-compose.yml [hunk 6]
```diff
@@ -120,7 +120,7 @@ services:
       HASH_GRAPH_PG_PORT: "5432"
       HASH_GRAPH_PG_DATABASE: "${HASH_GRAPH_PG_DEV_DATABASE}"
       HASH_GRAPH_LOG_FORMAT: "${HASH_GRAPH_LOG_FORMAT:-pretty}"
-      HASH_GRAPH_OTLP_ENDPOINT: "http://telemetry-collector:4317"
+      #      HASH_GRAPH_OTLP_ENDPOINT: "http://telemetry-collector:4317"
```
Comment (thehabbos007) [inline packages/hash/external-services/docker-compose.yml:123]: If we want to disable tracing in this PR, should we add an explainatory comment as in https://github.com/hashintel/hash/blob/fb9a93c0a37e61cbc4580247cf53e4f883c10488/packages/hash/external-services/docker-compose.yml#L123-L125
and maybe even link to the task we've created?

File: packages/hash/shared/src/util.ts [hunk 7]
```diff
@@ -233,3 +233,33 @@ export const isSafariBrowser = () =>
 export const typedKeys = <T extends keyof any>(
   record: Record<T, unknown>,
 ): T[] => Object.keys(record) as any;
+
+type TupleEntry<
+  T extends readonly unknown[],
+  I extends unknown[] = [],
+  R = never,
+> = T extends readonly [infer Head, ...infer Tail]
+  ? TupleEntry<Tail, [...I, unknown], R | [`${I["length"]}`, Head]>
```
Comment (thehabbos007) [inline packages/hash/shared/src/util.ts:242]: :panic-intensifies:

File: packages/hash/subgraph/src/types/element.ts [hunk 8]
```diff
@@ -35,7 +35,7 @@ export type EntityTypeWithMetadata = Omit<
 export type { OntologyElementMetadata } from "@hashintel/hash-graph-client";
 
 /** Plain JSON value and object definitions */
-type JsonValue = string | number | boolean | JsonObject | JsonValue[];
+type JsonValue = null | string | number | boolean | JsonObject | JsonValue[];
```
Comment (thehabbos007) [inline packages/hash/subgraph/src/types/element.ts:38]: I really wonder why there isn't a built in `JsonValue` type in TS..

**Review by thehabbos007 — COMMENTED** [review]
Very cool! I tested this locally and it seems to work fine ingesting data from GH. I did run into an issue which I've commented on regarding the catalog file :)

File: packages/hash/task-executor/src/main.ts [hunk 9]
```diff
@@ -1,5 +1,3 @@
-/* eslint-disable canonical/filename-no-index -- @todo rename file */
-
 import { readFileSync, writeFileSync } from "node:fs";
 
 import { json } from "body-parser";
```
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/hash/task-executor/src/main.ts:3]: Ah yeah I noticed this on the PR back in March (should have given the warning this time again, apologies), not sure it's worth the dev effort at the moment as we don't really want to keep storing them on the FS as artifacts anyway

File: packages/hash/api/src/graphql/resolvers/task-executor/generation.ts [hunk 10]
```diff
@@ -0,0 +1,495 @@
...
+const BOOLEAN_DATA_TYPE_ID =
+  "https://blockprotocol.org/@blockprotocol/types/data-type/boolean/v/1";
+const NULL_DATA_TYPE_ID =
+  "https://blockprotocol.org/@blockprotocol/types/data-type/null/v/1";
+const EMPTY_LIST_DATA_TYPE_ID =
+  "https://blockprotocol.org/@blockprotocol/types/data-type/empty-list/v/1";
+// const OBJECT_DATA_TYPE_ID =
+//   "https://blockprotocol.org/@blockprotocol/types/data-type/object/v/1";
+
+export type JsonValue =
```
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/task-executor/generation.ts:40]: I think we'll want to export it in one place, but I don't think this PR is the right time to do that haha.

I agree though, we do have an issue of creating helper types and such in places where they don't end up reused.

File: packages/hash/api/src/graphql/resolvers/task-executor/index.ts (hunk 4 above)
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/task-executor/index.ts:154]: I think it does ensure they exist, but we could throw an assertion `!` in, which would then break if we ever changed the middleware (and honestly if we change the middleware are we going to remember to go through the full code path and look for such assertions), or we could do an additional cheap check here.

I've wanted us to rework the User type for a while now though, and establish something like `UninitializedUser`
https://app.asana.com/0/1201095311341924/1202944961125764/f

File: packages/hash/subgraph/src/types/element.ts (hunk 8 above)
Comment (Alfred-Mountfield) [reply packages/hash/subgraph/src/types/element.ts:38]: Yep I have no idea 🙃

File: packages/hash/api/src/graphql/resolvers/task-executor/index.ts (hunk 5 above)
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/task-executor/index.ts:218]: image

Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/task-executor/index.ts:218]: Unless you're referring to something else? :)

File: packages/hash/external-services/docker-compose.yml (hunk 6 above)
Comment (Alfred-Mountfield) [reply packages/hash/external-services/docker-compose.yml]: Good callout, I've updated the comment, cheers

File: packages/hash/task-executor/src/main.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/hash/task-executor/src/main.ts:3]: I suppose that's fair. Should we at least add a bullet to the task executor readme to `mkdir -p src/tasks/source-github/secrets`?

File: packages/hash/api/src/graphql/resolvers/task-executor/index.ts (hunk 4 above)
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/task-executor/index.ts:154]: `UninitializedUser` seems like a good way to handle this. I don't think we need to change anything here now

**Review by thehabbos007 — DISMISSED** [review]
Looks good to merge! Thank you :)

File: packages/hash/task-executor/src/main.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/hash/task-executor/src/main.ts:3]: Good call out for maintainability, have added a quick comment but we don't even have a usage guide at the moment so I won't sink too much time in :)

**Review by thehabbos007 — DISMISSED** [review]

**Review by thehabbos007 — DISMISSED** [review]

**Review by thehabbos007 — APPROVED** [review]
~third~ fourth time is the charm
